PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,153 +3,217 @@
-    originalGraph = new OriginalGraph();
-    prepNodes(originalGraph, numNodes);
-    read();
-
-    sc.close();
-    sc = null;
+  public FullGfaReader(String fileName) {
+    read(fileName);
   }
 
   /**
-   * Initialize the nodes of a graph.
+   * Read a GFA file.
    *
-   * @param graph the graph of which to initialize a set amount of node.
-   * @param nodes the amount of nodes to initialize.
+   * @param fileName the name of the GFA file.
    */
-  protected static void prepNodes(OriginalGraph graph, int nodes) {
-    for (int i = 1; i < nodes + 1; i++) {
-      graph.addNode(new Node(i, 1, new ArrayList<>(), 0));
+  private void read(String fileName) {
+    try {
+      BufferedReader br = new BufferedReader(new InputStreamReader(FullGfaReader.class
+          .getClassLoader().getResourceAsStream(fileName)));
+      br.readLine();
+      String line;
+      while ((line = br.readLine()) != null) {
+        char[] chars = line.toCharArray();
+        switch (chars[0]) {
+          case 'L':
+            parseEdge(chars);
+            break;
+          case 'S':
+            parseNode(chars);
+            break;
+          case 'H':
+            parseHeader(chars);
+            break;
+          default:
+        }
+      }
+    } catch (IOException ex) {
+      Logger.getLogger(FullGfaReader.class.getName()).log(Level.SEVERE, null, ex);
     }
   }
-  
+
   /**
-   * Read the actual data.
+   * Parse a line of the GFA file which contains an edge.
+   *
+   * @param chars characters of the line.
    */
-  private void read() {
-    int count = 0;
-    while (sc.hasNext()) {
-      count++;
-      String next = sc.next();
-      switch (next) {
-        case "H":
-          handleH(sc);
-          break;
-        case "S":
-          handleS(sc);
-          break;
-        case "L":
-          handleL(sc);
-          break;
-        default:
-          System.out.println("Not S, H or L at: " + count);
-          sc.nextLine();
-      }
+  private void parseEdge(char[] chars) {
+    int from = 0;
+    int index = 2;
+    while (chars[index] != '\t') {
+      from *= 10;
+      from += chars[index++] - 48;
     }
-    System.out.println("Number of lines read: " + count);
+    index += 3;
+    int to = 0;
+    while (chars[index] != '\t') {
+      to *= 10;
+      to += chars[index++] - 48;
+    }
+    getNode(to).addInlink(from);
+    getNode(from).addOutlink(to);
   }
 
   /**
-   * Private method for handling an "H" as first character on the line of a gfa file.
-   * @param sc The scanner which is used to parse the file.
+   * Parse a line of the GFA file which contains a node.
+   *
+   * @param chars characters of the line.
    */
-  private void handleH(Scanner sc) {
-    String line = sc.nextLine();
-    this.genoms = extractGenomes(line);
+  private void parseNode(char[] chars) {
+    int nodeId = 0;
+    int index = 2;
+    while (chars[index] != '\t') {
+      nodeId *= 10;
+      nodeId += chars[index++] - 48;
+    }
```

Comment:
This loop reads a number from an array of characters. 48 is the ascii code of number 0 (I could change this to '0'). *10 because when you read one more number, the previously read number must be shifted by one in base 10.
It is annoying to create a separate method for this because the index has to be updated and the nodeId has to be remembered too. This can be solved by returning a value pair from a method which contains both the ID and the new index, but this is bad for performance (an extra method call, creating variables in the method, returning values in an object which has to be created and updating the variables in the method which called it). Usually this is not an issue, but this has to be high performance code and is performed many times (once for every node which is read and the other lines which you commented on is perfomed twice for every edges).

The magic numbers +2 / +3 / etc are dependent on the format of the file:
an edge line for example looks like this:
`L	2030	+	2031	+	0M`
to skip the 2 tabs and '+' in between the two numbers I for example use `index += 3;`. I don't think there is any way to describe the meaning of 3 in this case with a variable name, as it depends on the file format. This might seem like a hacky way, it is, but this is mainly caused by the file format which we get (binary would be much better) and we can't change that.
As long as the file format doesn't change we won't have to change this code, so imo screw coding standards and lets go for performace.